Vegetation Bending Tutorial (#1757)
o3de/o3de.org: content/docs/learning-guide/tutorials/rendering/vegetation-bending-tutorial.md

3. Add a call to our detail bending function under the wind function call in our `ProcessBending` function:
   
   ```
   if (o_color_isBound) {
      // Overall wind
      float4 currentBending = SetUpWindBending(currentTime, worldPosition);

      // Detail bending
      float3 currentOutPosition = DetailBending(position, normal, color, currentTime, worldPosition, currentBending.w);

      currentOutPosition = MainBending(currentOutPosition, currentBending);

      adjustedWorldPosition = mul(objectToWorld, float4(currentOutPosition, 1.0));
   }
   ```
5. Open the **Editor** and you should see your tree swaying with the leaves still bending. If you don't, try opening the **Material Editor** and increasing the wind parameters.

Amazing, our tree now sways and reacts to wind! Try to place multiple trees and observe how the trees sway differently when close together versus farther away. Also, add some lighting to make the trees pop!

{{< video src="/images/atom-guide/vegetation-bending-tutorial/tree3.mp4" autoplay="true" loop="true" width="100%" muted="true" info="Video of trees swaying." >}}

### Add motion vectors

> Because our tree is moving, we can add a *motion vector pass* so we can choose to add motion blur or other effects. For example, we have a partially implemented [Temporal Anti-Aliasing (TAA)](../../../atom-guide/features/taa.md) feature which uses the motion vectors. 

Reviewer: 'we can add' -> 'we should add'

Also, I would remove the 'partially implemented'. It is fully functional, there are just a lot of quality improvements we could make to it. But if they want to know about that, they can read more in the TAA document link you have right here.

If you take a look at the `MeshMotionVectorVegetationBending.azsl` file, you can see that the vertex shader looks similar to the other vertex shaders we have seen, but there's a new output `OUT.m_worldPosPrev`. This is the position of the vector in the last time frame. The pixel shader uses both the previous vector position and the current one to calculate the motion vector. 

> However, we don't have an vertex shader input that gives us the previous position. Therefore, in the vertex shader, we want to calculate the bending for not only the current time as we have been, but also for the previous time frame.

Reviewer: an vertex shader input -> a vertex shader input

3. Open `VegetationBending.materialtype`. At the bottom of the `shaders` list, add the motion vector pass:
   
   ```
   {
      "file": "./MeshMotionVectorVegetationBending.shader",
      "tag": "MeshMotionVector"
   }
   ```

Great, now we have included the motion vector shader, so we have to go in and edit the vertex shader.

1. Open `MeshMotionVectorVegetationBending.azsl`.
2. We want to perform bending on the vertex at the `IN.m_position` but at the previous frame time, with the previous world position. We can use `SceneSrg::m_prevTime` to get the previous frame time. Call our `ProcessBending` function again, but with the appropriate time and world position:
   ```
   float currentTime = SceneSrg::m_time;
   worldPosition = ProcessBending(currentTime, IN.m_position, IN.m_normal, IN.m_optional_color, worldPosition, objectToWorld);
   float prevTime = SceneSrg::m_prevTime;
   prevWorldPosition = ProcessBending(prevTime, IN.m_position, IN.m_normal, IN.m_optional_color, prevWorldPosition, prevObjectToWorld);
   ```

> 3. Take a look at the pixel shader to see how the motion vector is calculated! There is no need to edit the pixel shader.

Reviewer: Take a look at the pixel shader -> Take a look at the vertex shader.

Author: Not sure what you meant, do you mean to replace "Take a look at the pixel shader to see how the motion vector is calculated!" with "Take a look at the vertex shader."?

Reviewer: Leave it as is. I miss-understood...It makes sense this way in the context of having the whole complete file. Step 2 says edit the vertex shader, step 3 says Look at (but don't modify) the pixel shader.

Originally I was reading the first sentence of step 3 as if it were saying "look at the shader you just modified in step 2 to see how the motion vector is calculated", and thinking that there was something else going on in the vertex shader to calculate the motion vector after this snippit. But I can see now that there is nothing else afterwards in the vertex shader, and you really did mean to look at the pixel shader.

Author: Got it!

1. Move all the files in `atom_gems/AtomTutorials/Assets/VegetationBending/Objects/` to `{your_project_path}\Objects`. Make the `Objects` folder as needed.
1. Open `{your-project-path}\Materials\Types\VegetationBending.materialtype`. Under `propertyLayout` > `propertyGroups`, you'll see there are many entries with `{your-path-to-o3de}`. Replace `{your-path-to-o3de}` with your appropriate path to the engine.
   * For example, `C:/o3de/Gems/Atom/Feature/Common/Assets/Materials/Types/MaterialInputs/BaseColorPropertyGroup.json`.
   * Currently we cannot import property groups across gems, so we are hard coding the absolute path as a proof of concept, even though it is not portable. There is a GHI to enable importing across gems at [o3de#10623](https://github.com/o3de/o3de/issues/10623).
1. Run the **Editor**, and the assets should automatically process. At this point, none of them should fail. If `VegetationBending.materialtype` fails to process, check that you used the correct paths in step 4.

> These template files were created by duplicating important parts of the `StandardPBR` files and then modifying them. When you create your own material types in the future, you can similalry duplicate `StandardPBR` files and work from there.

Reviewer 2: This is really helpful information for those who want to develop their own material type. However, I don't think it fits in this section. At this point, the user is getting ready to do this tutorial, so information that's not specific to their current workflow at hand can be distracting and get forgotten. 

One way you might separate this information from the information relative to the tutorial is by putting it in a note callout at the end of the section. 
So, you can update to the following, and put it at the end of this section, (after "As a high-level overview..."):

```
{{< note > }}
These template files were created by duplicating important parts of the `StandardPBR` files and then modifying them. When you create your own material types in the future, you can similarly duplicate `StandardPBR` files and work from there.
{{< /note > }}
```

> As a high-level overview, [`.materialtype`](/docs/atom-guide/dev-guide/materials/materials/#material-types) references the shader files we will use on the material of this material type. The [`.shader`](/docs/atom-guide/look-dev/shaders/shader-file-spec.md) files define which types of shaders, such as vertex and pixel shaders, should be used and references the actual shader code in [`.azsl`](/docs/atom-guide/dev-guide/shaders/azsl/) files. They also specify the `DrawList`, which controls which pass should run that shader. Often, `.azsl` files will include `.azsli` files, which are also written in the Amazon shading language (AZSL), but are separate so multiple `.azsl` files can reuse the shader code from the `.azsli` files. 

Reviewer 2: Regarding previous suggestion to break up the Get Started section into the following subsections: 
- Get started
  - Prerequisites
  - Preparing your files
  - Concepts to know

I also suggest breaking up this paragraph to make it easier to read. 

```suggestion
As a high-level overview:
- The [`.materialtype`](/docs/atom-guide/dev-guide/materials/materials/#material-types) file references the shader files you will use on the material of this material type. 
- The [`.shader`](/docs/atom-guide/look-dev/shaders/shader-file-spec.md) files define which types of shaders, such as vertex and pixel shaders, should be used, and references the actual shader code in [`.azsl`](/docs/atom-guide/dev-guide/shaders/azsl/) files. They also specify the `DrawList`, which controls which pass should run that shader. 
- Often, `.azsl` files will include `.azsli` files, which are also written in the Amazon Shading Language (AZSL). These files are separate so multiple `.azsl` files can reuse the shader code from the `.azsli` files. 
```

## Add a material with the VegetationBending material type
Before we begin editing any files, we want to ensure we can make a material using our material type in the **Editor**.

>  1. Launch the **Editor**, and open the **Material Editor** by pressing **M** on the keyboard or choosing **Tools** > **Material Editor**. Give the **Material Editor** a few seconds to open.

Author: I went with your suggestion! I thought it might also be a good idea to link the Material Editor also since it's such a big part of the tutorial.

>  1. Create a new material by choosing **File** > **New**. Then in the dropdown under **Select Type**, choose `VegetationBending` and give the material a name, such as `my_material`. Save it somewhere in your project folder, such as in your project's `Materials` folder.

Author: Good point, the first "Save" in step 2 is actually just choosing the file location, so I updated that accordingly!

   {{< note >}}
   If you can't find the **Asset Processor**, click the arrow on the bottom right of your taskbar, and then click on the icon of two arrows in a circle.
   {{< /note >}}
1. Once the **Asset Processor** is done processing the changes, you should see in the **Editor** that your material looks different!

{{< image-width src="/images/atom-guide/vegetation-bending-tutorial/offset.png" width="100%" alt="Forward pass offset." >}}

The main texture of our shader ball seems to be showing up at an offset as we intended, but a grey outline is still at the origin of the object. This is because we only edited the forward pass, but have yet to edit the *depth pass*. All the passes we go through are referenced in `VegetationBending.materialtype`.
More information about passes can be found [here](/docs/atom-guide/dev-guide/passes/), but keep in mind that different passes render different parts of the material, and some passes' outputs are used as inputs to other passes.

Let's repeat the above steps but with the depth pass:
1. Open `{your-project-path}\Materials\Types\VegetationBending_DepthPass.azsli`.

>    * Make sure you are not editing the `VegetationBending_DepthPass_WithPS.azsli` file.

Author: Actually, this one is pretty hard to see but the difference is the `WithPS` suffix. Would you suggest I write this with a bigger emphasis on `WithPS`?

1. Find the function `DepthPassVS`.
1. Towards the end of the function, right before `OUT.m_position = mul(ViewSrg::m_viewProjectionMatrix, worldPosition);`, add: 
   
   ```
   worldPosition.x += 5.0;
   ```
1. Save your file and look at the **Editor**. The shader ball should now be completely rendered at an offset! 

>    * Note that the shadow is still in the original position. We'll be correcting this later on in the tutorial when we add a custom shadowmap with pixel shader. For brevity, we are not covering the vertex-only shadowmap shader in the tutorial and instead using the StandardPBR one. After finishing this tutorial, see if you can add and adjust the vertex-only shadow (used for opaque materials that don't have alpha-cutout) yourself!

Author: Hmm, these two sentences are actually supposed to refer to the current state, but I can see how it seems to relate to the shadowmap pass in the future section.
> For brevity, we are not covering the vertex-only shadowmap shader in the tutorial and instead using the StandardPBR one. After finishing this tutorial, see if you can add and adjust the vertex-only shadow (used for opaque materials that don't have alpha-cutout) yourself!

I changed this paragraph slightly to hopefully make more sense in this place! Let me know if I should make any further changes.

   ```
   {
      "$import": "MaterialInputs/VegetationBendingPropertyGroup.json"
   },
   ```
   Note that you can also place the `json` properties directly in this `.materialtype` file, without having to import another `.json` file. 

Great, now we have parameters, but let's actually use the parameters in the code and view them.
1. Open `{your-project-path}\Materials\Types\VegetationBending_ForwardPass.azsli`. 
1. We can reference the x offset parameter by using `MaterialSrg::m_xOffset`. So, where you previously put `worldPosition.x += 5.0`, replace it and add the y offset: 

   ```
   worldPosition.x += MaterialSrg::m_xOffset;
   worldPosition.y += MaterialSrg::m_yOffset;
   ```
   Recall that we added the parameters into the material shader resource group, so that's how we can reference them with MaterialSrg.
1. Repeat with the depth pass.
1. Save your files and open the **Material Editor**.
1. Select the material you made with the VegetationBending material type previously (`my_material`), and find **Vegetation Bending** in the **Inspector** on the right. Adjust the x and y offsets as you see fit!
1. Save and return to the **Editor**.

> 1. Observe how the offset matches your inputs from the **Material Editor**!

Reviewer 2: ```suggestion
1. Observe how the offset matches your inputs from the Material Editor!
```

Now, we have a tree (at an offset)! This tree is important because it uses vertex colors that we will be using to determine how the tree should bend.

Note on vertex colors: You can use a DCC tool to color the vertices to match the type of bending you want to do on each part of the tree: 
* Red vertex color: smaller movement with high frequency of random sinusoidal noise.
* Green vertex color: delays the start of the movement for variations with high frequency of random sinusoidal noise.
* Blue vertex color: larger movement and bending with low frequency of sinusoidal noise. 
In the case of our tree, the trunk's vertices are colored blue and leaves are colored red.

### Add a shader option
Our tree mesh that we just added has colored vertices; however, not every mesh that you want to apply your material type to may have colored vertices. Therefore, we want to add a [*shader option*](/docs/atom-guide/dev-guide/shaders/azsl/#shader-variant-options) that allows the vertex shader to handle both of these cases. We also want to add the reference to the colors so we can use them in the code.

1. Open `{your-project-path}\Materials\Types\VegetationBending_Common.azsli`.

> 1. At the bottom of the file, add:

Reviewer 2: - Consolidate steps
- This suggestion uses the file name only. But more importantly, you should be consistent. So only do this if you're also updating the file paths to use the file name only (per my earlier comment). 

```suggestion
1. In the `VegetationBending_Common.azsli` file, at the bottom, define a boolean shader option:
```

   ```
   option bool o_color_isBound;
   ```
   We place this variable in the common file because we will be using it in all the passes.

> 1. Open `{your-project-path}\Materials\Types\VegetationBending_ForwardPass.azsli`. Inside `struct VegetationVSInput`, add another field:

Reviewer 2: Same comment about file names here. 
```suggestion
1. In the `VegetationBending_ForwardPass.azsli` file, inside `struct VegetationVSInput`, add the following field. 
```

   ```
   float4 m_optional_color : COLOR0;
   ```

>    If the material's vertices are colored, `m_optional_color` will be set at runtime if it's available. Then, if `m_optional_color` is available, a soft naming convention will set `o_color_isBound` to true, which we can use later to determine if we want to perform the bending or not. Note that this soft naming convention is a very specific sub-feature of shader options that are set based on the presence or absence of an optional vertex stream, as opposed to shader options set based on material properties.

Reviewer 2: Some revisions. Also, consider distinguishing this set of conceptual information.

Also, I assume you meant "mesh's vertices", not "material's vertices"?
```suggestion
   For a mesh with colored vertices, `m_optional_color` gets set at runtime, if it's available. Then, a soft name convention sets `o_color_isBound` to `true`, which we can use to determine if we want to perform the bending or not. 

   *Soft naming convention* is a very specific sub-feature of a shader option. It's set based on the presence or absence of an optional 
```

>    All of the fields are indicated by [HLSL semantics](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-semantics#vertex-shader-semantics). The engine processes the semantics and updates the fields accordingly.

Reviewer 2: Similarly, consider wrapping this in a note callout.

> 1. Inside the function `VegetationBending_ForwardPassVS`, encase the offset code with a conditional using the option boolean:

Reviewer 2: ```suggestion
1. Inside the function `VegetationBending_ForwardPassVS`, encase the offset code with an `if`-condition by using the boolean shader option:
```

   ```
   if (o_color_isBound) {
      worldPosition.x += MaterialSrg::m_xOffset;
      worldPosition.y += MaterialSrg::m_yOffset;
   }
   ```

> 1. Repeat steps 3 and 4 with the depth pass in `{your-project-path}\Materials\Types\VegetationBending_DepthPass.azsli`.

Reviewer 2: If you consolidated the first two steps, then update this: 
```suggestion
1. Repeat steps 2 and 3 with the depth pass in `{your-project-path}\Materials\Types\VegetationBending_DepthPass.azsli`.
```

> 1. Save both files and open your level in the **Editor** from the previous steps. You should see that the tree entity is offset, but the shader ball (which doesn't have a vertex color stream) is not!

Reviewer 2: ```suggestion
1. Save both files and open your level in the **Editor** from the previous steps. 

You should see that the tree entity is offset, but the shader ball is not! That's because the shader ball doesn't have a vertex color stream.
```

{{< image-width src="/images/atom-guide/vegetation-bending-tutorial/optionoffset.png" width="100%" alt="Option offset added." >}}

### Delete previous offset code

> Now that we have our tree and tested the shader options, let's delete the previous offset code so that we can start fresh with vegetation bending. That is, delete the following:

Reviewer 2: ```suggestion
Let's delete the previous offset code so that we can implement vegetation bending. 

Delete the following: 
```

> * The code added to adjust the position in the vertex shaders of both the forward pass and depth pass
> * The `m_xOffset` and `m_yOffset` variable declarations in the `VegetationBending_Common.azsli` file in the `MaterialSrg`
> * The `xOffset` and `yOffset` properties in the `VegetationBendingPropertyGroup.json` file, but keep the rest of the file and its connection in `VegetationBending.materialtype`; we will need it for later!

Reviewer 2: ```suggestion
* The code added to adjust the position in the vertex shaders of both the forward pass and depth pass.
* The `m_xOffset` and `m_yOffset` variable declarations in the `MaterialSrg`, in the `VegetationBending_Common.azsli` file.
* The `xOffset` and `yOffset` properties in the `VegetationBendingPropertyGroup.json` file. However,  keep the rest of the file and its connection in `VegetationBending.materialtype`.
```

> Make sure you are keeping the declarations for `m_optional_color` and `o_color_isBound`.

Reviewer 2: ```suggestion
Make sure to keep the declarations for `m_optional_color` and `o_color_isBound`.
```

### Create materials for the tree

> Let's add some textures to make our tree look more realistic! For the tree, we need 3 materials: one for the trunk, one for the branches, and one for the leaves.

Reviewer 2: ```suggestion
Let's add some textures to make our tree look more realistic! For the tree, we need 3 materials: for the trunk, the branches, and the leaves.
```

> 

Reviewer 2: For the steps in this section, I suggest revising to better streamline the instructions. It's best to put the goal of the step near the beginning. 
You can refer back to my comments for **Add a tree** section for examples.

Author: Made some changes to this section with putting the goal at the beginning of the step and shortening the actions!

1. Open the Editor, and then the **Material Editor**.

> 1. Choose **File** > **New** and, in the pop-up, choose **VegetationBending**, name the material `aspen_leaf.material`, and save it in the same folder you saved your previous material in, such as the `Materials` folder.

Reviewer 2: You can simplify this significantly, since you already taught how to create a material in **Add a material with the VegetationBending material type**. 

The most important parts of this step is to create a new material of the VegetationBending material type named `aspen_leaf.material`, and to save it in the `Materials` folder.

1. On the right side in the **Inspector**, find **Base Color** and click on the file icon next to *Texture*. Choose `aspen_leaf_basecolora.tif`.

>    * The suffix `_basecolora` tells the engine to process the texture with a specific [*preset*](/docs/user-guide/assets/texture-settings/texture-presets). Appending a suffix to the name of a texture tells the engine to use the corresponding preset. In this case, we are using the `_basecolora` preset because this texture has the base color in the rgb channels and the opacity in the alpha channel.

Reviewer 2: Distinguish this concept from the step, such as by using a callout.

1. Find **Opacity** and, for **Opacity Mode**, select `Cutout`. We need to set this to `Cutout` because the leaf texture has transparent parts.
1. Under **General Settings**, enable **Double-sided**. This renders both sides of meshes.
1. Save your leaf material. Now, we need to repeat with the branch and the trunk. Instead of making whole new materials, we can make the leaf material the parent of the branch and trunk materials so the properties stay constant for all 3.
   1. In the **Asset Browser** of the Material Editor, **right-click** `aspen_leaf.material`. Select **Create Child Material...** and save it as `aspen_bark_01.material` in the same folder as `aspen_leaf.material`. Find **Base Color** in the **Inspector** and choose `aspen_bark_01_basecolor.tif`.

>    1. In the **Asset Browser** of the Material Editor, **right-click** `aspen_leaf.material`. Select **Create Child Material...** and save it as `aspen_bark_02.material` in the same folder as `aspen_leaf.material`. Find **Base Color** in the **Inspector** and choose `aspen_bark_02_basecolor.tif`.

Reviewer 2: Consider simplifying this to repeat the above step for `aspen_bark_02.material`

>    Notice how the other properties are the same as the leaf's! If you edit the parent material's properties after creating these child materials, it will automatically update the child materials' property values. This will be important later when we adjust the bending properties so all parts of the tree remain in sync while bending.

Reviewer 2: Distinguish this concept from the step, such as by using a callout.

1. Save all 3 materials and exit the Material Editor. In the **Editor**, click on the tree entity (`Tree`).

> 1. In the **Entity Inspector** on the right, find the **Material** component. Click the **X** to delete the **Default Material**. Under **Model Materials** > **AM_Aspen_Bark_01**, select `aspen_bark_01.material`. For **AM_Aspen_Bark_02**, select `aspen_bark_02.material`. For **AM_Aspen_Leaf**, select `aspen_leaf.material`.

Reviewer 2: You can organize the substeps here into a list for better readability. For example: 

> For the **Model Materials** property, map the following materials: 
> - **AM_Aspen_Bark_01**: `aspen_bark_01.material`
> - **AM_Aspen_Bark_02**: `aspen_bark_02.material`
> - **AM_Aspen_Leaf**: `aspen_leaf.material`

## Add vegetation bending
Great, now we can start adding the code for vegetation bending! We first need to set up the wind constants. Then, we will determine the detail bending, which is the slight movement that you generally see in leaves and at the end of branches. Finally, we will add main bending, which is the overall swaying of the tree.

Note that the following bending functions are derived from [Vegetation Procedural Animation and Shading in Crysis](https://developer.nvidia.com/gpugems/gpugems3/part-iii-rendering/chapter-16-vegetation-procedural-animation-and-shading-crysis) in GPU Gems 3.

### Add vegetation bending parameters
We need several parameters to determine how we want our materials to bend:
* DetailBendingFrequency - The frequency of the detail bending.
* DetailBendingLeafAmplitude - The amplitude in which leaves can bend.
* DetailBendingBranchAmplitude - The amplitude in which branches can bend.
* WindX -The amount of wind in the x direction.
* WindY - The amount of wind in the y direction.
* WindBendingStrength - The amount in which the vegetation bends as a result of the wind.

> * WindBendingFrequency - The frequency that the object sways back and forth caused by the wind.

Reviewer 2: Use `inline code style` around the parameters

> The "DetailBending" parameters are specfically used for detail bending, but the "Wind" parameters are used for all parts of the bending.

Reviewer 2: ```suggestion
The `DetailBending`- parameters are specifically used for detail bending, while the `Wind`- parameters are used for all parts of the bending.
```

1. Open `VegetationBendingPropertyGroup.json`.

> 1. Delete the `xOffset` and `yOffset` properties that we added previously if you haven't already, and add these seven:

Reviewer 2: Shorten long sentences, especially if they contain multiple steps. 
```suggestion
1. Delete the `xOffset` and `yOffset` properties that we added previously, if you haven't already. Add these seven:
```

   
   ```hlsl
   float currentTime = SceneSrg::m_time;
   worldPosition = ProcessBending(currentTime, IN.m_position, IN.m_normal, IN.m_optional_color, worldPosition, objectToWorld);

   OUT.m_worldPosition = worldPosition.xyz;
   OUT.m_position = mul(ViewSrg::m_viewProjectionMatrix, worldPosition);

   return OUT;
   ``` 
   {{< note >}}
   We need the code to be above the two `OUT` lines because we will edit `worldPosition` and then will need to adjust the `OUT` variables accordingly.
   {{< /note >}}
  
1. Repeat steps 2-3 with the depth pass in `VegetationBending_DepthPass.azsli` and the depth pass with PS in `VegetationBending_DepthPass_WithPS.azsli`.

   {{< tip >}}
   Note that the depth pass does not output a world position. Just place the code above the `OUT.m_position` line. The depth pass with PS has some extra code in the vertex shader, but you can still place this code about the two `OUT` statements.
   {{< /tip >}}

### Set up wind bending
Let's begin editing the code to add wind. 

1. Open `VegetationBending_Common.azsli`.

> 1. At the bottom, add a function to calculate the amplitude, frequency, and phase of the wind according to the time and world position of the vertex. The phase should take into account the `worldPosition` because we want to mimic how wind should affect nearby objects similarly but faraway objects differently. Faraway objects may not be affected by the same "breeze". This function will then later be used to calculate the appropriate movement of the vertex.

Reviewer 2: Add new lines to break apart this step

```suggestion
1. At the bottom, add a function to calculate the amplitude, frequency, and phase of the wind according to the time and world position of the vertex. The wind's phase uses the `worldPosition` to mimic how wind affects nearby objects similarly, but faraway objects differently. This is because in real life, faraway objects may not be affected by the same breeze. 

  Later, we'll use this function to calculate the appropriate movement of the vertex.
```

   ```hlsl
   float4 SetUpWindBending(float currentTime, float4 worldPosition) 
   {
      float2 wind = float2(MaterialSrg::m_windX, MaterialSrg::m_windY);
      float bendingStrength = MaterialSrg::m_bendingStrength;
      float2 amplitude = float2(wind.x * 0.4 + wind.y * 0.2, wind.y * 0.4 - wind.x * 0.2);
      float2 frequency = float2(MaterialSrg::m_windBendingFrequency, MaterialSrg::m_windBendingFrequency * 1.125);
      // Using the world position to modify the phase makes it so different trees near each other are at similar but not equal points in the animation, 
      // so they appear to be reacting to the same wind but at different times as the wind moves through the vegetation.
      float2 phase = worldPosition.xy * 0.08;

      float2 bendAmount = sin(currentTime * frequency + phase) * amplitude;

      float4 result;
      result.xy = bendAmount + wind;
      result.z = length(wind);
      result.w = 0.3 * length(result.xy);
      result.xyz *= bendingStrength * 0.08;

      return result;
   }
   ```

   {{< note >}}

>    This function could potentially be run once per-object on the CPU with the results updated each frame in the ObjectSrg instead of being run once per-vertex on the GPU. It's a tradeoff between doing the computation once per frame per vegetation object + an extra SRG compile per frame per vegetation object vs. re-doing the computation on the GPU per frame per vegetation object per vertex. It's content specific, with the redundant GPU cost increasing as vertex density increases, and also depends on if the GPU is the bottleneck, and whether or not the vertex shader is the bottleneck, and also if the vertex shader is bandwidth bound or ALU bound. 

Reviewer 2: Can you do the following:
- expand the full term for ALU, then put ALU in parenthesis
- break apart the long sentences to avoid run-ons
- break apart into shorter paragraphs

```suggestion
   This function could potentially be run once per-object on the CPU with the results updated each frame in the ObjectSrg instead of being run once per-vertex on the GPU. It's a tradeoff between doing the computation once per frame per vegetation object, plus an extra SRG compile per frame per vegetation object versus re-doing the computation on the GPU per frame per vegetation object per vertex. It's content specific, with the redundant GPU cost increasing as vertex density increases, and also depends on if the GPU is the bottleneck, and whether or not the vertex shader is the bottleneck, and also if the vertex shader is bandwidth bound or ALU bound. 
```

      float2 wavesIn = currentTime;
      wavesIn += float2(vtxPhase, branchPhase);
      float4 waves = (frac(wavesIn.xxyy * float4(1.975, 0.793, 0.375,  0.193)) * 2.0 - 1.0) * MaterialSrg::m_detailFrequency * bendLength;
      waves = abs(frac(waves + 0.5) * 2.0 - 1.0);

      // x: is used for leaves, y is used for branches
      float2 wavesSum = waves.xz + waves.yw;

      // Leaf and branch bending (xy is used for leaves, z for branches)
      float3 movement = wavesSum.xxy * float3(edgeInfo * MaterialSrg::m_detailLeafAmplitude * normal.xy, branchBendAmount * MaterialSrg::m_detailBranchAmplitude);
      return objectSpacePosition + movement;
   }
   ```

1. Call the detail bending function in our `ProcessBending` function. Inside the conditional, add the call:
   
   ```hlsl
   if (o_color_isBound) 
   {
      // Overall wind
      float4 currentBending = SetUpWindBending(currentTime, worldPosition);

      // Detail bending
      float3 currentOutPosition = DetailBending(objectSpacePosition, normal, detailBendingParams, currentTime, worldPosition, currentBending.w);
   }

   return adjustedWorldPosition;
   ```

   {{< note >}}

>    Note that `currentBending.w` is passed in to the detail bending function. This is the overall bending length we want according to the wind strength and direction. 

Reviewer 2: ```suggestion
   The `currentBending.w` parameter that's passed into the `DetailBending` function controls the overall bending length according to the wind's strength and direction. 
```

   ```hlsl
   float currentTime = SceneSrg::m_time;
   worldPosition = ProcessBending(currentTime, IN.m_position, IN.m_normal, IN.m_optional_color, worldPosition, objectToWorld);
   ```

1. For our motion vectors to work we need to find the previous world position. We want to perform bending on the vertex at the `IN.m_position` but at the previous frame time, with the previous world position. We can use `SceneSrg::m_prevTime` to get the previous frame time. Call our `ProcessBending` function again, but with the appropriate time and world position:

   ```hlsl
   float currentTime = SceneSrg::m_time;
   worldPosition = ProcessBending(currentTime, IN.m_position, IN.m_normal, IN.m_optional_color, worldPosition, objectToWorld);
   float prevTime = SceneSrg::m_prevTime;
   prevWorldPosition = ProcessBending(prevTime, IN.m_position, IN.m_normal, IN.m_optional_color, prevWorldPosition, prevObjectToWorld);
   ```

1. Take a look at the pixel shader to see how the motion vector is calculated! There is no need to edit the pixel shader.

> Amazing, we have added everything we need to add for motion vectors! However, if you open the **Editor** and just view the tree, you'll see that there is no difference. We can, however, observe that the motion vector pass works by using the **pass tree visualizer**.

Reviewer 2: Space out the steps below by adding a new line between each step.

1. Open the Editor and press **Ctrl-G** to enter gameplay mode.
1. Press the **Home** key on your keyboard. This brings up the toolbar at the top.
1. Select **Atom Tools** > **Pass Viewer**.
1. In the pop-up **PassTree View**, enable **Preview Attachment** and **Show Pass Attachments**.
1. In the **PassTree**, find *MotionVectorPass* > *MeshMotionVectorPass* and select the line with `CameraMotion`. 

> 1. Ensure you are viewing your tree. Note that you probably won't see much except black on the bottom left preview, because the motion vectors are very small since our tree only moves minimally. However, if you move the camera around or translate the tree quickly, you may see some motion vectors pop up. You can also open `MeshMotionVectorVegetationBending.azsl` and scale `OUT.m_motion` in the pixel shader to ensure that the motion vectors' directions are working properly.

Reviewer 2: For long steps like this, consider breaking it up by adding a new line in between the shorter paragraphs. You can indent the paragraphs so they are aligned with the step. The space helps with readability. 

Some of this information may be better in a note callout box as well.

If you'd like to download and enable the **AtomTutorials Gem**, do the following:
1. Download or clone the [o3de/sample-code-gems repository](https://github.com/o3de/sample-code-gems). 

   {{< note >}}
   If you followed this tutorial, you already downloaded or cloned this repository, but probably moved the `Object` files out of the repository. You can move the files back or re-download/clone the repository.
   {{< /note >}}

1. Open `VegetationBending.materialtype` and replace all the instances of `{your-path-to-o3de}` with your absolute path to O3DE. 

1. [Register the **AtomTutorials Gem** to the engine](/docs/user-guide/project-config/register-gems). In the command line interface, `cd` into the engine. Then, run

   ```
   scripts\o3de register -gp {your-path-to-sample-code-gems}\atom_gems\AtomTutorials -espp {your-project-path}
   ```

   For example, `scripts\o3de register -gp C:\sample-code-gems\atom_gems\AtomTutorials -espp C:\MyProject`

> 1. [Add the Gem into your project](/docs/user-guide/project-config/add-remove-gems). 
> 
>    1. Open the **Project Manager**, which should be located in your project's `build/windows/bin/profile/o3de.exe`.
> 
>    1. In the menu, find your project, click on the three bars, and select **Configure Gems...**.
> 
>    1. In the **Gem Catalog**, search for **AtomTutorials** and enable the Gem.
> 
>    1. Click **Save** in the bottom right.

Reviewer 2: Since the linked page explains all of this, you can remove them here and consolidate into single step. 

```suggestion
1. Add the **AtomTutorials** Gem to you project. Follow the instructions in [Adding and Removing Gems in a Project](/docs/user-guide/project-config/add-remove-gems). 
```

> 1. Re-build your project by clicking **Build Project**, or on the three bars again and selecting **Build**. 

Reviewer 2: ```suggestion
1. Re-build your project by clicking **Build Project**, or by clicking {{< icon "menu.svg" >}} and selecting **Build**. 
```

1. Open `VegetationBending.materialtype`.

1. Find `shaders`, a list of the `.shader` files that your material type can use. Notice that each shader file is referenced with a tag. The tag allows us to reference the shader easily.

1. Under the shader with `"tag": "Shadowmap"`,  add a new entry for the shadowmap with PS:

   ```json
   {
      "file": "./VegetationBending_Shadowmap_WithPS.shader",
      "tag": "Shadowmap_WithPS"
   },
   ```

1. Similarly, under the shader with `"tag": "DepthPass"`,  add a new entry for the depth map with PS:

   ```json
   {
      "file": "./VegetationBending_DepthPass_WithPS.shader",
      "tag": "DepthPass_WithPS"
   },
   ```

1. Save the file.
   
You've added the appropriate shaders to the list of shaders for your material type. However, you may notice that just adding to the list of shaders doesn't change the tree. You will need to give the engine instructions for which shader to use for different materials. This is where `.lua` material functors come in. 

1. Open `VegetationBending_ShaderEnable.lua`.

> 1. Look through the file, and observe how it enables the pixel shader versions of the depth and shadowmap passes if `OpacityMode_Cutout` is used or if parallax with pixel depth offset is enabled. There is no need to edit anything in the `.lua` file for now.

Reviewer 2: ```suggestion
1. In the file, observe how it enables the pixel shader versions of the depth and shadowmap passes (`depthPassWithPS` and `shadowMapWithPS`) if parallax with pixel depth offset is enabled, or if `OpacityMode_Cutout` is used.
   
   There is no need to edit anything in this file for now.
```

1. Open `VegetationBending.materialtype`.

> 1. Include the `.lua` file so the engine can process it and determine which shaders to use. After the list for `"shaders"` and before `"uvNameMap"`, add:

Reviewer 2: ```suggestion
1. Create a functor and include the `VegetationBending_ShaderEnable.lua` file. This allows the engine to process it and determine which shaders to use. 
```

   ```json
      "functors": [
         {
            "type": "Lua",
            "args": {
               "file": "Materials/Types/VegetationBending_ShaderEnable.lua"
            }
         }
      ],
   ```

> 1. Save the file, allow the **Asset Processor** to process the changes, and open the **Editor** again. Observe how the tree now looks pretty realistic! 

Reviewer 2: ```suggestion
1. Save the file, allow the **Asset Processor** to process the changes, and open the **Editor** again. 

Observe how the tree looks more realistic! 
```

{{< image-width src="/images/atom-guide/vegetation-bending-tutorial/treeallpasses.png" width="100%" alt="The tree properly rendered in the Editor with all appropriate passes added." >}}

## Add vegetation bending

> Great, now you can start adding the code for vegetation bending! You first need to set up the wind constants. Then, you will determine the detail bending, which is the slight movement that you generally see in leaves and at the end of branches. Finally, you will add main bending, which is the overall swaying of the tree.

Reviewer 2: ```suggestion
Great, now you can start adding the code for vegetation bending! 

First, you need to set up the wind constants. Then, you will determine the detail bending, which is the slight movement that you see in leaves and at the end of branches. Finally, you will add main bending, which is the overall swaying of the tree.
```

> Note that the following bending functions are derived from [Vegetation Procedural Animation and Shading in Crysis](https://developer.nvidia.com/gpugems/gpugems3/part-iii-rendering/chapter-16-vegetation-procedural-animation-and-shading-crysis) in GPU Gems 3.

Reviewer 2: ```suggestion
{{< note >}}
The following bending functions are derived from [Vegetation Procedural Animation and Shading in Crysis](https://developer.nvidia.com/gpugems/gpugems3/part-iii-rendering/chapter-16-vegetation-procedural-animation-and-shading-crysis) in NVIDIA GPU Gems 3.
{{< /note >}}
```

         "type": "Float",
         "defaultValue": 0.0,
         "min": 0.0,
         "max": 7.0,
         "connection": {
         "type": "ShaderInput",
         "name": "m_bendingStrength"
         }
      },
      {
         "name": "WindBendingFrequency",
         "displayName": "Wind Bending Frequency",
         "description": "The frequency that the object sways back and forth caused by the wind. This would typically come from a wind system instead of a material property, but is here as a proof of concept.",
         "type": "Float",
         "defaultValue": 0.0,
         "min": 0.0,
         "max": 1.5,
         "connection": {
         "type": "ShaderInput",
         "name": "m_windBendingFrequency"
         }
      }
   ]
   }

   ```

1. Open `VegetationBending_Common.azsli`.

1. Delete the previous offset variables if you haven't already, and declare the bending property variables in `MaterialSrg`:
   
   ```hlsl
    float m_detailFrequency;
    float m_detailLeafAmplitude;
    float m_detailBranchAmplitude;
    float m_windX;
    float m_windY;
    float m_bendingStrength;
    float m_windBendingFrequency;
    ```

> 1. Open the **Editor**, then open the **Material Editor**, and choose to edit the leaf material (`aspen_leaf.material`). Make sure you select the leaf material because that is the parent material of the other parts of the tree.

Reviewer 2: ```suggestion
1. Open the leaf material (`aspen_leaf.material`) in the **Material Editor**. Make sure you select the leaf material because that is the parent material of the other parts of the tree.
```

> 1. On the right, scroll down to **Vegetation Bending** and ensure that the 6 properties you just added are there.

Reviewer 2: ```suggestion
1. In the **Inspector**, scroll to the **Vegetation Bending** property group. Ensure that the six properties you just added are there.
```

> 1. Adjust the properties! Make sure you adjust all of them so that you can see bending in the later steps.

Reviewer 2: Are there any recommendations on what to adjust the properties to?

Author: Changed this step to the following:

> 1. Adjust the properties! You can adjust them to the following to ensure you can see bending later:
>    * **Detail bending frequency** - `0.3`
>    * **Detail bending leaf amplitude** - `0.3`
>    * **Detail bending branch amplitude** - `0.3`
>    * **Wind direction x** - `0.5`
>    * **Wind direction y** - `0.5`
>    * **Bending strength** - `4.0`
>    * **Wind bending frequency** - `0.7`

---
linkTitle: Vertex Deformation for Vegetation Bending Tutorial
title: Vertex Deformation for Vegetation Bending Tutorial

> description: A tutorial for add vegetation bending with the Atom rendered in the Open 3D Engine (O3DE).

Reviewer 2: Put in "material type" and "shaders" since they're likely key search terms. Helps with SEO so people can find this page better. 
```suggestion
description: A tutorial for adding vegetation bending by creating custom a material type and authoring shaders with the Atom Renderer in the Open 3D Engine (O3DE).
```

toc: true
---

In this tutorial, we cover how to make your own material type and how to edit vertex shaders to achieve a vegetation bending effect. While we use vegetation bending as an example, the primary goal of this tutorial is to familiarize yourself with how to use and create custom material types and vertex shaders. 

> This tutorial covers the following concepts:
> * Making your own material type
> * Creating materials in the **Material Editor**
> * Editing vertex shaders
> * Adding editable properties for your material type
> * Using optional vertex streams
> * Adding passes to your material type
> * Using the **Pass Tree Visualizer** to debug passes

Reviewer 2: Reorganized the concepts so we can see how each is connected better

```suggestion
This tutorial covers the following concepts:
* Making your own material type
* Creating materials in the **Material Editor**
* Editing vertex shaders
* Adding editable properties for your material type
* Using optional vertex streams
* Adding passes to your material type
* Using the **Pass Tree Visualizer** to debug passes
```

Author: I understand what you mean about "using shader options," so I updated the docs to put a little more emphasis on "vertex streams." I kept the concept as "Using optional vertex streams" and renamed the section to "Use an optional vertex stream" because the shader option that I was adding was only a part of adding the vertex stream -- the vertex color is part of the vertex stream and I only use the shader option from whether the vertex stream is available, so it seems more appropriate to use "optional vertex stream." 

Otherwise, I took your suggestions for italicizing and indenting some concepts!

1. Open the Editor, and then the **Material Editor**.

1. Create a new material of the **VegetationBending** material type named `aspen_leaf.material`, and save it in the same folder you saved your previous material in, such as the `Materials` folder.

1. Set the base color texture of the material. In the **Inspector**, under the **Base Color** > **Texture** property, click {{< icon "file-folder.svg" >}} and choose `aspen_leaf_basecolora.tif`.

   {{< note >}}
   The suffix `_basecolora` tells the engine to process the texture with a specific [*preset*](/docs/user-guide/assets/texture-settings/texture-presets). Appending a suffix to the name of a texture tells the engine to use the corresponding preset. In this case, we are using the `_basecolora` preset because this texture has the base color in the rgb channels and the opacity in the alpha channel.
   {{< /note >}}

1. Set the opacity mode. For the **Opacity** > **Opacity Mode** property, select `Cutout`. We need to set this to `Cutout` because the leaf texture has transparent parts.

1. Under **General Settings**, enable **Double-sided**. This renders both sides of the material.

> 1. Save your leaf material. Repeat with the branch and the trunk, but instead of making whole new materials, make the leaf material the parent of the branch and trunk materials so the properties stay constant for all 3.

Reviewer 2: - Split into two steps. 
- "Repeat" step was getting long, so I suggest splitting the sentences.
- Added other terms to avoid ambiguity. 
```suggestion
1. Save your `aspen_leaf.material` material. 
2. Repeat the above steps for the branch and the trunk materials, but don't make new materials. Instead, make the branch and trunk materials be children of the leaf material. This allows the material properties to stay constant for all three.
```

>    1. In the **Asset Browser** of the Material Editor, **right-click** `aspen_leaf.material`. Select **Create Child Material...** and save it as `aspen_bark_01.material` in the same folder as `aspen_leaf.material`. Find **Base Color** in the **Inspector** and choose `aspen_bark_01_basecolor.tif`.

Reviewer 2: ```suggestion
   1. In the **Asset Browser** of the Material Editor, **right-click** `aspen_leaf.material`. 
      - Select **Create Child Material...** and save it as `aspen_bark_01.material` in the same folder as `aspen_leaf.material`. 
      - Find **Base Color** in the **Inspector** and choose `aspen_bark_01_basecolor.tif`.
```

The main texture of the shader ball shows up at an offset as intended, but a grey outline is still at the origin of the object. This is because you only edited the forward pass, and have yet to edit the *depth pass*. All the passes this material goes through are referenced in `VegetationBending.materialtype`.
Keep in mind that different passes render different parts of the material, and some passes' outputs are used as inputs to other passes. You can find more information about passes in the [Passes](/docs/atom-guide/dev-guide/passes/) section.

Repeat the above steps for the depth pass:
1. Open `VegetationBending_DepthPass.azsli`.

   {{< caution >}}
   Make sure you are not editing the `VegetationBending_DepthPass_WithPS.azsli` file.
   {{< /caution >}}

1. Find the function `DepthPassVS`.

1. Towards the end of the function, right before `OUT.m_position = mul(ViewSrg::m_viewProjectionMatrix, worldPosition);`, add: 
   
   ```
   worldPosition.x += 5.0;
   ```

1. Save your file and look at the Editor. The shader ball should now be completely rendered at an offset! 
   
   {{< note >}}

>    Note that the shadow is still in the original position. This is because this tutorial doesn't cover the vertex-only shadowmap shader and instead using the StandardPBR one for brevity. You will be adding a custom shadowmap with pixel shader later on in the tutorial. However, after finishing this tutorial, see if you can add and adjust the vertex-only shadow (used for opaque materials that don't have alpha-cutout) yourself! 

Reviewer 2: I'm confused by the mention of the "vertex-only" shadowmap shader. I'm not sure if it's needed here. Similarly, I think that a "vertex-only shadow (used for opaque materials that don't have alpha-cutout)" needs more context for the suggestion to adjust it to be actionable. 

In retrospect, using the term "vertex-only" may make sense as a way to identify something. But right now and at this part of the doc, it's not clear to me what you're referring to. 

I think the core information to call out here is: 
>  Note that the shadow is still in the original position. That's because you haven't updated the shadowmap shader, yet. Later in the tutorial, you will add a custom shadowmap with a pixel shader, which will fix the shadow.

My suggestion below is only to rephrase, doesn't include anything above.
```suggestion
   Note that the shadow is still in the original position. This is because this tutorial doesn't cover the vertex-only shadowmap shader and instead uses the StandardPBR one for brevity. Later in the tutorial, you will add a custom shadowmap with a pixel shader. However, after finishing this tutorial, see if you can add and adjust the vertex-only shadowmap shader yourself! Recall that the vertex-only shadowmap shader is used for opaque materials that don't have alpha-cutout.
```

Author: I think you're right, I'm introducing some new terminology that at this point the user probably wouldn't understand (or care about) yet, so I went ahead and just put in what you noted as the important information.

1. In the Material Editor **Inspector**, find **Vegetation Bending** and ensure that the 6 properties you just added are there.

1. Adjust the properties! Make sure you adjust all of them so that you can see bending in the later steps.

### Add process bending function
First, let's add a function that your multiple vertex shaders can call to avoid repeat code. You will be writing more functions for different parts of the bending and calling them all from this function. 
1. Open `VegetationBending_Common.azsli`.

1. At the bottom, add a function to return the world position of the vertex after applying bending. The parameters given to this function are all helpful for determining bending.

   ```hlsl
   float4 ProcessBending(float currentTime, float3 objectSpacePosition, float3 normal, float4 detailBendingParams, float4 worldPosition, float4x4 objectToWorld) 
   {
      float4 adjustedWorldPosition = float4(worldPosition);
      if (o_color_isBound) 
      {
         // You will add function calls here.
      }
      return adjustedWorldPosition;
   }
   ```

   {{< note >}}

>    Notice how, like before, you are using a conditional with `o_color_isBound` to ensure that only meshes with vertex streams perform bending.

Reviewer 2: ```suggestion
   Like before, notice how you use a conditional with `o_color_isBound` to ensure that only meshes with vertex streams perform bending.
```

   {{< /note >}}

> 1. Call the `ProcessBending` function in your vertex shaders. Open `VegetationBending_ForwardPass.azsli` and find the vertex shader (`VegetationBending_ForwardPassVS`). Add a call to your function above `OUT.m_worldPosition = worldPosition.xyz`. Most of the parameters to pass in are inputs to your vertex shader or values you have calculated already, and you can get the time from the property `m_time` the *scene shader resource group* (`SceneSrg`), which gives us the number of seconds since the start of the application.

Reviewer 2: ```suggestion
1. Call the `ProcessBending` function in your vertex shaders.

   1. Open `VegetationBending_ForwardPass.azsli` and find the vertex shader, `VegetationBending_ForwardPassVS`. 
   
   1. Above `OUT.m_worldPosition = worldPosition.xyz`, call the `ProcessBending` function. 
   
      The parameters to pass in are inputs to your vertex shader, values you have calculated already, and `m_time`, the number of seconds since the start of the application. `m_time` is provided by the *scene Shader Resource Group* (`SceneSrg`).
```

   
   ```hlsl
   float currentTime = SceneSrg::m_time;
   worldPosition = ProcessBending(currentTime, IN.m_position, IN.m_normal, IN.m_optional_color, worldPosition, objectToWorld);

   OUT.m_worldPosition = worldPosition.xyz;
   OUT.m_position = mul(ViewSrg::m_viewProjectionMatrix, worldPosition);

   return OUT;
   ``` 
   {{< note >}}

>    The code needs to be above the two `OUT` lines because you will edit `worldPosition`, which adjusts the `OUT` variables accordingly.

Reviewer 2: ```suggestion
   The code must be above the two `OUT` lines because it updates the `worldPosition`, which adjusts the `OUT` variables accordingly.
```

   {{< /note >}}
  
1. Repeat steps 2-3 with the depth pass in `VegetationBending_DepthPass.azsli` and the depth pass with PS in `VegetationBending_DepthPass_WithPS.azsli`.

   {{< tip >}}

>    Note that the depth pass does not output a world position. Just place the code above the `OUT.m_position` line. The depth pass with PS has some extra code in the vertex shader, but you can still place this code about the two `OUT` statements.

Reviewer 2: If there's no special instructions for the depth pass with PS, I suggest removing it completely.
```suggestion
   For  `VegetationBending_DepthPass.azsli`, you only need to place the code above the `OUT.m_position` line because the depth pass does not output a world position
```

   
   ```glsl
   if (o_color_isBound)
   {
      // Overall wind
      float4 currentBending = SetUpWindBending(currentTime, worldPosition);

      // Detail bending
      float3 currentOutPosition = DetailBending(position, normal, color, currentTime, worldPosition, currentBending.w);

      currentOutPosition = MainBending(currentOutPosition, currentBending);

      adjustedWorldPosition = mul(objectToWorld, float4(currentOutPosition, 1.0));
   }
   ```

1. Open the **Editor**. You should see that your tree sways and the leaves still bend. If you don't, try increasing the wind properties in the Material Editor.

Amazing, your tree now sways and reacts to wind! Try to place multiple trees and observe how the trees sway differently when close together versus farther away. Also, add some lighting to make the trees pop!

{{< video src="/images/learning-guide/tutorials/rendering/vegetation-bending-tutorial/tree3.mp4" autoplay="true" loop="true" width="100%" muted="true" info="Three trees in the Editor, all swaying at a different pace with detail bending." >}}

### Add motion vectors

> Because your tree is moving, you should add a *motion vector pass* so you can choose to add motion blur or other effects. For example, we have a [Temporal Anti-Aliasing (TAA)](/docs/atom-guide/features/taa.md) feature which uses the motion vectors.

Reviewer 2: ```suggestion
Since your tree moves, you can add cool effects by using a *motion vector pass*. Motion vectors are used by effects such as motion blur and [Temporal Anti-Aliasing (TAA)](/docs/atom-guide/features/taa.md).
```

> If you take a look at the `MeshMotionVectorVegetationBending.azsl` file, you can see that the vertex shader looks similar to the other vertex shaders you have seen, but there's a new output `OUT.m_worldPosPrev`. This is the position of the vector in the last time frame. The pixel shader uses both the previous vector position and the current one to calculate the motion vector. 

Reviewer 2: ```suggestion
If you look at the `MeshMotionVectorVegetationBending.azsl` file, the vertex shader looks similar to the other vertex shaders, but there's a new output. `OUT.m_worldPosPrev` is the position of the vector in the previous frame. The pixel shader uses both the previous vector position and the current one to calculate the motion vector. 
```

   ```

1. For motion vectors to work you need to find the previous world position. You will want to perform bending on the vertex at the `IN.m_position` but at the previous frame time, with the previous world position. You can use `SceneSrg::m_prevTime` to get the previous frame time. Call your `ProcessBending` function again, but with the appropriate time and world position:

   ```glsl
   float currentTime = SceneSrg::m_time;
   worldPosition = ProcessBending(currentTime, IN.m_position, IN.m_normal, IN.m_optional_color, worldPosition, objectToWorld);
   float prevTime = SceneSrg::m_prevTime;
   prevWorldPosition = ProcessBending(prevTime, IN.m_position, IN.m_normal, IN.m_optional_color, prevWorldPosition, prevObjectToWorld);
   ```

1. Take a look at the pixel shader to see how the motion vector is calculated! There is no need to edit the pixel shader.

Amazing, you have added everything you need to add for motion vectors! However, if you open the **Editor** and just view the tree, you'll see that there is no difference. You can, however, observe that the motion vector pass works by using the **pass tree visualizer**.

> 1. Open the Editor and press **Ctrl-G** to enter gameplay mode.

Reviewer 2: ```suggestion
#### Debugging with the Pass Tree Visualizer

1. Open the Editor and press **Ctrl-G** to enter gameplay mode.
```

1. Press the **Home** key on your keyboard. This brings up the toolbar at the top.

1. Select **Atom Tools** > **Pass Viewer**.

1. In the pop-up **PassTree View**, enable **Preview Attachment** and **Show Pass Attachments**.

1. In the **PassTree**, find *MotionVectorPass* > *MeshMotionVectorPass* and select the line with `CameraMotion`. 

> 1. Ensure you are viewing your tree. You probably won't see much except black on the bottom left preview, because the motion vectors are very small since the tree only moves minimally. 
> 
>    However, if you move the camera around or translate the tree quickly, you may see some motion vectors pop up. You can also open `MeshMotionVectorVegetationBending.azsl` and scale `OUT.m_motion` in the pixel shader to ensure that the motion vectors' directions are working properly.

Reviewer 2: ```suggestion
1. Ensure you are viewing your tree. 
   
   {{< note >}}
   You may only see black on the bottom left preview, because the motion vectors are small, meaning the tree moves minimally. 

   To better see the motion vectors, you can move the camera around or translate the tree quickly. You can also open `MeshMotionVectorVegetationBending.azsl` and scale `OUT.m_motion` in the pixel shader to ensure that the motion vectors' directions are working properly.
   {{< /note >}}
```

This video shows the motion vectors when `OUT.m_motion` is scaled by `10000.0`.
{{< video src="/images/learning-guide/tutorials/rendering/vegetation-bending-tutorial/motionvectortree.mp4" autoplay="true" loop="true" muted="true" width="100%" info="Three trees in the Editor swaying with detail bending, with the motion vector visualizer on the bottom right indicating the direction of movement." >}}

> The pass tree visualizer tool is also helpful with debugging shaders and passes, so you can see the output of certain steps of different passes when you select them in the **PassTree**.

Reviewer 2: Cool tool! Will definitely use it next time.
```suggestion
{{< tip >}}
The Pass Tree Visualizer tool is also helpful for debugging shaders and passes. It allows you to see the output of certain steps of different passes when you select them in the **PassTree**.
{{< /tip >}}
```

This tutorial covers the following concepts:
* Creating *materials* in the **Material Editor**
* Creating a custom *material type*
  * Adding *adjustable properties* for your material type
  * Adding *passes* to your material type
* Editing *vertex shaders*
  * Using *optional vertex streams*
* Using the **Pass Tree Visualizer** to debug passes

The **VegetationBending** material type allows materials to bend and sway, simulating how wind affects vegetation. It allows for detail bending with slight movement of branches and leaves, as well as movement of the entire object. 

The code in this tutorial can be found in the [**Atom Tutorials Gem**](https://github.com/o3de/sample-code-gems/tree/main/atom_gems/AtomTutorials) in the `o3de/samples-code-gems` repository. There, you can find the template code and assets needed for this tutorial, as well as the final code.

As you go along, you can reference the [Material Types and Shaders](get-started-materialtypes-and-shaders) tutorial, which gives higher-level explanations of the mechanisms used in this tutorial.

## Get started

> Before you can create your material type, ensure you have [installed the engine](/docs/welcome-guide/setup/), [set up a project](/docs/welcome-guide/create/), and [launched the editor](/docs/welcome-guide/tours/editor-tour).

Reviewer 2: ```suggestion
Before you can create your material type, ensure you have [installed the engine](/docs/welcome-guide/setup/), [set up a project](/docs/welcome-guide/create/), and [launched the Editor](/docs/welcome-guide/tours/editor-tour).
```

{{< /tip >}}

## Download the AtomTutorial Gem sample
Now that you've completed this tutorial, you can compare your results to our working version of the vegetation bending material type in the **AtomTutorials Gem** in the [o3de/sample-code-gems repository](https://github.com/o3de/sample-code-gems). You can either download and place the [final working vegetation bending files](https://github.com/o3de/sample-code-gems/tree/main/atom_gems/AtomTutorials/Assets/VegetationBending) in your project, or you can download the Gem and add it to the engine.

If you'd like to download and enable the **AtomTutorials Gem**, do the following:
1. Download or clone the [o3de/sample-code-gems repository](https://github.com/o3de/sample-code-gems). 

   {{< note >}}
   If you followed this tutorial, you already downloaded or cloned this repository, but probably moved the `Object` files out of the repository. You can move the files back or re-download/clone the repository.
   {{< /note >}}

1. Open `VegetationBending.materialtype` and replace all the instances of `{your-path-to-o3de}` with your absolute path to O3DE. 

> 1. [Register the **AtomTutorials Gem** to the engine](/docs/user-guide/project-config/register-gems). In the command line interface, `cd` into the engine. Then, run
> 
>    ```
>    scripts\o3de register -gp {your-path-to-sample-code-gems}\atom_gems\AtomTutorials -espp {your-project-path}
>    ```
> 
>    For example, `scripts\o3de register -gp C:\sample-code-gems\atom_gems\AtomTutorials -espp C:\MyProject`.

Reviewer 2: I suggest removing the link because you already provide the command that the user needs to run. The link might confuse readers into thinking they need to complete the steps listed there, even though it's already here. 
````suggestion
1. Register the **AtomTutorials Gem** to your project. In the command line interface, run the following command:

   ```
   cd {path-to-o3de-engine}
   scripts\o3de register -gp {your-path-to-sample-code-gems}\atom_gems\AtomTutorials -espp {your-project-path}
   ```

   For example, with paths:
   ```
   scripts\o3de register -gp C:\sample-code-gems\atom_gems\AtomTutorials -espp C:\MyProject
   ```
````

  Later, you'll use this function to calculate the appropriate movement of the vertex.

   ```hlsl
   float4 SetUpWindBending(float currentTime, float4 worldPosition) 
   {
      float2 wind = float2(MaterialSrg::m_windX, MaterialSrg::m_windY);
      float bendingStrength = MaterialSrg::m_bendingStrength;
      float2 amplitude = float2(wind.x * 0.4 + wind.y * 0.2, wind.y * 0.4 - wind.x * 0.2);
      float2 frequency = float2(MaterialSrg::m_windBendingFrequency, MaterialSrg::m_windBendingFrequency * 1.125);
      // Using the world position to modify the phase makes it so different trees near each other are at similar but not equal points in the animation, 
      // so they appear to be reacting to the same wind but at different times as the wind moves through the vegetation.
      float2 phase = worldPosition.xy * 0.08;

      float2 bendAmount = sin(currentTime * frequency + phase) * amplitude;

      float4 result;
      result.xy = bendAmount + wind;
      result.z = length(wind);
      result.w = 0.3 * length(result.xy);
      result.xyz *= bendingStrength * 0.08;

      return result;
   }
   ```

   {{< note >}}

>    This function could potentially be run once per-object on the CPU with the results updated each frame in the ObjectSrg instead of being run once per-vertex on the GPU. It's a tradeoff between doing the computation with an extra SRG compile once per frame per vegetation object versus re-doing the computation on the GPU per frame per vegetation object per vertex.
>    
>    The choice is content specific, with the redundant GPU cost increasing as vertex density increases. The choice also depends on if the GPU is the bottleneck, if the vertex shader is the bottleneck, and if the vertex shader is bandwidth bound or arithmetic logic unit (ALU) bound. 

Reviewer 2: ```suggestion
   By default, this function runs once per vertex on the GPU. Instead, you can potentially run it once per object on the CPU, causing the results to update each frame in the ObjectSrg. The tradeoff is between recomputing on the GPU per object, per vertex, per frame, versus computing with an extra SRG compile once per frame, per object.

   Your choice may depend on how much content you have, since the redundant GPU cost increases as vertex density increases. Your choice may also depend on whether the GPU or the vertex shader is the bottleneck, or if the vertex shader is bandwidth bound or arithmetic logic unit (ALU) bound. 
```

   Your choice may depend on how much content you have, since the redundant GPU cost increases as vertex density increases. Your choice may also depend on whether the GPU or the vertex shader is the bottleneck, or if the vertex shader is bandwidth bound or arithmetic logic unit (ALU) bound. 
   {{< /note >}}

2. Call the function in your `ProcessBending` function. Inside the conditional, add a call to your wind function:
   
   ```glsl
   if (o_color_isBound) 
   {
      // Overall wind
      float4 currentBending = SetUpWindBending(currentTime, worldPosition);
   }

   return adjustedWorldPosition;
   ```

Great, now you have your wind bending function set up! Note that this doesn't enact any changes on the tree just yet, and the tree should be rendered as normal.

### Add detail bending
Using the wind bending constants that you just calculated, you can now determine the bending of the leaves.

1. Open `VegetationBending_Common.azsli`.

> 1. Above your `ProcessBending` function, add a function that calculates the amount of movement and returns the resulting position for a vertex.

Reviewer 2: ```suggestion
1. Add a `DetailBending` function that calculates the amount of movement and returns the resulting position for a vertex. Place this above the `ProcessBending` function.
```

   {
      // The information from the vertex colors about how to bend this vertex.
      float edgeInfo = detailBendingParams.x;
      float branchPhase = detailBendingParams.y;
      float branchBendAmount = detailBendingParams.z;

      // Phases (object, vertex, branch)
      float objPhase = dot(worldPosition.xyz, 2.0); 
      branchPhase += objPhase;
      float vtxPhase = dot(objectSpacePosition, branchPhase); 

      // Detail bending for leaves
      // x: is used for leaves, y is used for branch
      float2 wavesIn = currentTime;
      wavesIn += float2(vtxPhase, branchPhase);
      float4 waves = (frac(wavesIn.xxyy * float4(1.975, 0.793, 0.375,  0.193)) * 2.0 - 1.0) * MaterialSrg::m_detailFrequency * bendLength;
      waves = abs(frac(waves + 0.5) * 2.0 - 1.0);

      // x: is used for leaves, y is used for branches
      float2 wavesSum = waves.xz + waves.yw;

      // Leaf and branch bending (xy is used for leaves, z for branches)
      float3 movement = wavesSum.xxy * float3(edgeInfo * MaterialSrg::m_detailLeafAmplitude * normal.xy, branchBendAmount * MaterialSrg::m_detailBranchAmplitude);
      return objectSpacePosition + movement;
   }
   ```

> 1. Call the detail bending function in your `ProcessBending` function. Inside the conditional, add the call:
>    
>    ```glsl
>    if (o_color_isBound) 
>    {
>       // Overall wind
>       float4 currentBending = SetUpWindBending(currentTime, worldPosition);
> 
>       // Detail bending
>       float3 currentOutPosition = DetailBending(objectSpacePosition, normal, detailBendingParams, currentTime, worldPosition, currentBending.w);
>    }
> 
>    return adjustedWorldPosition;
>    ```
> 
>    {{< note >}}
>    The `currentBending.w` parameter that's passed into the `DetailBending` function controls the overall bending length according to the wind's strength and direction. 
>    {{< /note >}}
> 
> 1. You will need to set the actual vertex shader outputs to use the output from the detail bending function. Set the world position so the code following the conditional can update the positions correctly. 
>    
>    ```glsl
>    if (o_color_isBound) 
>    {
>       // Overall wind
>       float4 currentBending = SetUpWindBending(currentTime, worldPosition);
> 
>       // Detail bending
>       float3 currentOutPosition = DetailBending(position, normal, detailBendingParams, currentTime, worldPosition, currentBending.w);
> 
>       adjustedWorldPosition = mul(objectToWorld, float4(currentOutPosition, 1.0));
>    }
> 
>    return adjustedWorldPosition;
>    ```

Reviewer 2: Since both of these steps happen inside the conditional, I suggest consolidating them and having only one code snippet. I also tried to rephrase the second step - please verify if this is still accurate. 

The suggestion below would look like: 
![image](https://user-images.githubusercontent.com/75444793/182721793-c11b4966-63e0-43d3-b9e2-2ab9deb2d612.png)

````suggestion
1. In the `ProcessBending` function, inside the conditional:

   * Call the `DetailBending` function.
   
   * Set and return the adjusted world position, so the actual vertex shader output uses the output from the `DetailBending` function. 
   
   ```hlsl
   if (o_color_isBound) 
   {
      // Overall wind
      float4 currentBending = SetUpWindBending(currentTime, worldPosition);

      // Detail bending
      float3 currentOutPosition = DetailBending(position, normal, detailBendingParams, currentTime, worldPosition, currentBending.w);

      adjustedWorldPosition = mul(objectToWorld, float4(currentOutPosition, 1.0));
   }

   return adjustedWorldPosition;
   ```
   
   {{< note >}}
   The `currentBending.w` parameter that's passed into the `DetailBending` function controls the overall bending length according to the wind's strength and direction. 
   {{< /note >}}
````

> 1. Open the Editor and you should see your tree's leaves bending slightly. If you don't, try opening the Material Editor and increasing all the properties. 

Reviewer 2: ```suggestion
1. Open the Editor. You should see that your tree's leaves bend slightly. If you don't, try increasing all the properties in the Material Editor.
```

{{< video src="/images/learning-guide/tutorials/rendering/vegetation-bending-tutorial/detailbendingtree.mp4" autoplay="true" loop="true" width="100%" muted="true" info="The tree in the Editor with detail bending applied, moving the leaves slightly." >}}

### Add main bending

> The leaves are moving now, but the tree doesn't sway yet. You will now add main bending, the overall sway and movement that the whole tree experiences.

Reviewer 2: ```suggestion
The leaves move now, but the tree doesn't sway yet. In this step, you will add main bending, the overall sway and movement that the whole tree experiences.
```

1. Open `VegetationBending_Common.azsli`.

1. Above your `ProcessBending` function, add a function to make the tree sway. Using the current position of the vertex (after it has been changed from the detail bending) and the bending determined by the wind, you can bend the tree as a whole.
   
   ```glsl
   float3 MainBending(float3 objectSpacePosition, float4 bending)
   {
      float windX = bending.x;
      float windY = bending.y;
      float bendScale = bending.z;

      // More bending occurs higher up on the object
      float bendFactor = objectSpacePosition.z * bendScale;
      bendFactor *= bendFactor; 

      // Rescale the displaced vertex position with the original distance to the object's center
      // to restrict vertex movement to minimize deformation artifacts
      float len = length(objectSpacePosition); 
      float3 newPos = objectSpacePosition;
      newPos.xy += float2(windX, windY) * bendFactor;

      return normalize(newPos) * len;
   }
   ```

> 1. Add a call to your main bending function under the detail bending function call in your `ProcessBending` function:

Reviewer 2: ```suggestion
1. Call the `MainBending` function after the call to the `DetailBending` function, in the `ProcessBending` function.
```

   
   ```glsl
   if (o_color_isBound)
   {
      // Overall wind
      float4 currentBending = SetUpWindBending(currentTime, worldPosition);

      // Detail bending
      float3 currentOutPosition = DetailBending(position, normal, color, currentTime, worldPosition, currentBending.w);

      currentOutPosition = MainBending(currentOutPosition, currentBending);

      adjustedWorldPosition = mul(objectToWorld, float4(currentOutPosition, 1.0));
   }
   ```

> 1. Open the **Editor** and you should see your tree swaying with the leaves still bending. If you don't, try opening the **Material Editor** and increasing the wind properties.

Reviewer 2: ```suggestion
1. Open the **Editor**. You should see that your tree sways and the leaves still bend. If you don't, try increasing the wind properties in the Material Editor.
```

If you take a look at the `MeshMotionVectorVegetationBending.azsl` file, you can see that the vertex shader looks similar to the other vertex shaders you have seen, but there's a new output `OUT.m_worldPosPrev`. This is the position of the vector in the last time frame. The pixel shader uses both the previous vector position and the current one to calculate the motion vector. 

However, there isn't a vertex shader input that gives us the previous position. Therefore, in the vertex shader, you will calculate the bending for not only the current time as you have been, but also for the previous time frame.

Let's add the motion vector shader and then edit its vertex shader:

1. Open `VegetationBending.materialtype`. At the bottom of the `shaders` list, add the motion vector pass:
   
   ```json
   {
      "file": "./MeshMotionVectorVegetationBending.shader",
      "tag": "MeshMotionVector"
   }
   ```

1. Open `MeshMotionVectorVegetationBending.azsl`.

> 1. Add the call to `ProcessBending`, just as you did in the previous steps with the other shaders. Under the declaration for `float4 prevWorldPosition`, but above `OUT.m_worldPos`, add:
> 
>    ```glsl
>    float currentTime = SceneSrg::m_time;
>    worldPosition = ProcessBending(currentTime, IN.m_position, IN.m_normal, IN.m_optional_color, worldPosition, objectToWorld);
>    ```
> 
> 1. For motion vectors to work you need to find the previous world position. You will want to perform bending on the vertex at the `IN.m_position` but at the previous frame time, with the previous world position. You can use `SceneSrg::m_prevTime` to get the previous frame time. Call your `ProcessBending` function again, but with the appropriate time and world position:
> 
>    ```glsl
>    float currentTime = SceneSrg::m_time;
>    worldPosition = ProcessBending(currentTime, IN.m_position, IN.m_normal, IN.m_optional_color, worldPosition, objectToWorld);
>    float prevTime = SceneSrg::m_prevTime;
>    prevWorldPosition = ProcessBending(prevTime, IN.m_position, IN.m_normal, IN.m_optional_color, prevWorldPosition, prevObjectToWorld);
>    ```

Reviewer 2: I suggest consolidating this step into a single step with substeps. I think it better wraps the idea of how motion vectors work. I also simplified the second step, regarding previous  frame times / world positions. This also let's you use one code block. 

Suggestions below results in this: 
![image](https://user-images.githubusercontent.com/75444793/182727427-1f327214-b255-47eb-b73a-2bb7ad746f43.png)

````suggestion
1. For motion vectors to work, you need to perform bending at the current frame time and the previous frame time.
 
   Under the `float4 prevWorldPosition` declaration, above `OUT.m_worldPos`, add the following:

   1. Call the `ProcessBending` function and pass in the current time and world position. This is similar to the calls you made in the earlier shaders.
   
   1. Call `ProcessBending` again, but this time, pass in the previous frame time and previous world position. Use `SceneSrg::m_prevTime` to get the previous frame time.

   ```glsl
   float currentTime = SceneSrg::m_time;
   worldPosition = ProcessBending(currentTime, IN.m_position, IN.m_normal, IN.m_optional_color, worldPosition, objectToWorld);
   
   float prevTime = SceneSrg::m_prevTime;
   prevWorldPosition = ProcessBending(prevTime, IN.m_position, IN.m_normal, IN.m_optional_color, prevWorldPosition, prevObjectToWorld);
   ```
````

1. Take a look at the pixel shader to see how the motion vector is calculated! There is no need to edit the pixel shader.

> Amazing, you have added everything you need to add for motion vectors! However, if you open the **Editor** and just view the tree, you'll see that there is no difference. You can, however, observe that the motion vector pass works by using the **pass tree visualizer**.

Reviewer 2: ```suggestion
Amazing, you have added everything you need to add for motion vectors! However, if you open the **Editor** and just view the tree, you'll see that there is no difference. To observe that the motion vector pass works, you can use the **Pass Tree Visualizer**.
```

---

> linkTitle: Vertex Deformation for Vegetation Bending Tutorial

Author: This is a good suggestion, but I renamed the link title to be "Vertex Deformation" instead since that's the main concept of the tutorial.

1. The template files for this tutorial are in `atom_gems/AtomTutorials/Templates/VegetationBending/`. Move all of the files to `{your-project-path}\Materials\Types\`.

   {{< tip >}}
   Notice that there are a couple of significant files and file extensions that you need to make your custom material type. This tutorial will explain each file as you go along. You can also find more details in the [Material Types and Shaders](get-started-materialtypes-and-shaders) tutorial. 
   {{< /tip >}}

1. Move all the files in `atom_gems/AtomTutorials/Assets/VegetationBending/Objects/` to `{your_project_path}\Objects`. Make the `Objects` folder as needed.

1. Open `{your-project-path}\Materials\Types\VegetationBending.materialtype` in a text editor. 

1. Under `propertyLayout` > `propertyGroups`, notice that there are many entries with `{your-path-to-o3de}`. Replace `{your-path-to-o3de}` with the appropriate path to your engine.  

   For example, the resulting path might look like: `C:/o3de/Gems/Atom/Feature/Common/Assets/Materials/Types/MaterialInputs/BaseColorPropertyGroup.json`.
   

>    {{< note >}}
>    Currently, O3DE cannot import property groups across Gems. So, you must hard code the absolute path as a proof of concept, even though hard-coding is not recommended as it restricts portability. 
>    
>    There is a [GitHub issue](https://github.com/o3de/o3de/issues/10623) to enable importing across Gems.
>    {{< /note >}}

Reviewer 2: ```suggestion
   {{< known-issue >}}
   Currently, O3DE cannot import property groups across Gems. So, you must hard code the absolute path as a proof of concept, even though hard-coding is not recommended as it restricts portability. 
   
   There is a [GitHub issue](https://github.com/o3de/o3de/issues/10623) to enable importing across Gems.
   {{< /known-issue>}}
```

- Often, `.azsl` files will include `.azsli` files, which are also written in the Amazon Shading Language (AZSL). These files are separate so multiple `.azsl` files can reuse the shader code from the `.azsli` files. 

{{< tip >}}
These template files were created by duplicating important parts of the `StandardPBR` files and then modifying them. When you create your own material types in the future, you can similarly duplicate `StandardPBR` files and work from there.
{{< /tip >}}

## Add a material with the VegetationBending material type
Before you begin editing any files, make a material using your material type in the **Editor**.

 1. Launch the Editor, and open the **Material Editor**. For information on how to do this, see the [Material Editor](/docs/atom-guide/look-dev/materials/material-editor) page.

 1. Create a new material by choosing **File** > **New**. Then in the **Select Type** drop down, choose **VegetationBending** and give the material a name, such as `my_material`. Choose the file location to be somewhere in your project folder, such as in your project's `Materials` folder.

>  1. Save your material by hitting **Ctrl-S**. Then, close the Material Editor.

Reviewer 2: ```suggestion
 1. Save your material by pressing **Ctrl-S**. Then, close the Material Editor.
```

 1. Back in the Editor, select the *shader ball* that is already included in the default level. The **Entity Inspector** should now show the properties of the shader ball object. 

 1. In the Entity Inspector, look for the **Mesh** component of the shader ball and click **Add Material Component**.

 1. In the Material component of the shader ball, click the file icon next to *Default Material*. Then, select the VegetationBending material, named `my_material`, that you just created.

{{< image-width src="/images/learning-guide/tutorials/rendering/vegetation-bending-tutorial/material.png" width="100%" alt="Adding a VegetationBending material to an object's Material component in the O3DE Editor." >}}

> Great, you just created a material with your custom material type!

Reviewer 2: ```suggestion
Great, you created a material with your custom material type!
```

You should see that the tree entity is offset, but the shader ball is not! That's because the shader ball doesn't have a vertex color stream.

{{< image-width src="/images/learning-guide/tutorials/rendering/vegetation-bending-tutorial/optionoffset.png" width="100%" alt="The tree and the shader ball in the Editor, with only the tree offset from adding the shader option." >}}

### Delete previous offset code
Delete the previous offset code so that you can implement vegetation bending. 

Delete the following: 
* The code added to adjust the position in the vertex shaders of both the forward pass and depth pass.
* The `m_xOffset` and `m_yOffset` variable declarations in the `MaterialSrg`, in the `VegetationBending_Common.azsli` file.
* The `xOffset` and `yOffset` properties in the `VegetationBendingPropertyGroup.json` file. However,  keep the rest of the file and its connection in `VegetationBending.materialtype`.
* In the Editor, `my_material` from the **Default Material** in the **Material** component for both the tree and the shader ball.

Make sure to keep the declarations for `m_optional_color` and `o_color_isBound`.

### Create materials for the tree

> Add some textures to make your tree look more realistic! For the tree, you need 3 materials: for the trunk, the branches, and the leaves.

Reviewer 2: ```suggestion
Add some textures to make your tree look more realistic! For the tree, you need three materials: for the trunk, the branches, and the leaves.
```

1. Save your `aspen_leaf.material` material. 
2. Repeat the above steps for the branch and the trunk materials, but don't make new materials. Instead, make the branch and trunk materials be children of the leaf material. This allows the material properties to stay constant for all three.

   1. In the **Asset Browser** of the Material Editor, **right-click** `aspen_leaf.material`. 
      - Select **Create Child Material...** and save it as `aspen_bark_01.material` in the same folder as `aspen_leaf.material`. 
      - Find **Base Color** in the **Inspector** and choose `aspen_bark_01_basecolor.tif`.

   1. Repeat the above step for `aspen_bark_02.material`. Make it a child of `aspen_leaf.material` and set **Base Color** as `aspen_bark_02_basecolor.tif`.

   {{< note >}}
   Notice how the other properties are the same as the leaf's! If you edit the parent material's properties after creating these child materials, it will automatically update the child materials' property values. This will be important later when you adjust the bending properties so all parts of the tree remain in sync while bending.
   {{< /note >}}

> 1. Save all 3 materials and exit the Material Editor. In the **Editor**, select the tree entity (`Tree`).

Reviewer 2: ```suggestion
1. Save all three materials and exit the Material Editor. In the **Editor**, select the tree entity (`Tree`).
```

1. Towards the end of the function, right before `OUT.m_worldPosition = worldPosition.xyz;`, add: 
   
   ```hlsl
   worldPosition.x += 5.0;
   ```
   This will adjust the position in the positive x direction by `5` units. You may wonder why we are editing `worldPosition` instead of `m_position`; `m_position` is the position of this vertex relative to the origin of the model, whereas `m_worldPosition` is the position of this vertex relative to the origin of the level (or world). Try out editing the other dimensions and `m_position` and `worldPosition` and see what they do!
1. Make sure the **Editor** is open, if it is not already open.
1. Save your file with **CTRL-S** and the **Asset Processor** should automatically detect changes and process the file. You can open the **Asset Processor** and check when the file is done processing. 
   {{< note >}}
   If you can't find the **Asset Processor**, click the arrow on the bottom right of your taskbar, and then click on the icon of two arrows in a circle.
   {{< /note >}}
1. Once the **Asset Processor** is done processing the changes, you should see in the **Editor** that your material looks different!
   {{< note >}}

>    The Editor viewport will not update unless the Editor window is in focus.

Reviewer 3: This is no longer true, @gadams3 fixed it to reload even when not in focus.

Author: Thanks, fixed

## Download the AtomTutorial Gem sample
Now that you've completed this tutorial, you can compare your results to our working version of the vegetation bending material type in the **AtomTutorials** gem in the [o3de/sample-code-gems repository](https://github.com/o3de/sample-code-gems). You can either download and place the [final working vegetation bending files]() in your project, or you can download the gem and add it to the engine.

If you'd like to download and enable the **AtomTutorials** gem, do the following:
1. Download or clone the [o3de/sample-code-gems repository](https://github.com/o3de/sample-code-gems). 
1. Open `VegetationBending.materialtype` and replace all the instances of `{your-path-to-o3de}` with your absolute path to O3DE. 
1. [Register the **AtomTutorials** gem to the engine](/docs/user-guide/project-config/register-gems). In the command line interface, `cd` into the engine. Then, run
   ```
   scripts\o3de register -gp {your-path-to-sample-code-gems}\atom_gems\AtomTutorials -espp {your-project-path}
   ```

>    For example, `scripts\o3de register -gp C:\sample-code-gems\atom_gems\AtomTutorials -espp C:\MyProject`

Reviewer 3: I think you can do this step through o3de.exe too, I remember seeing an Add Gem button.

Author: Hmm, I can find adding the gem but I can't seem to find an option to register a new gem. The docs (https://www.o3de.org/docs/user-guide/project-config/register-gems/) also don't mention using o3de.exe, so I thought it'd be safer to just provide the CLI commands.
